### PR TITLE
Add support for IAddCarry SPIRV instruction

### PR DESCRIPTION
**Pull request by bwlodarcz**
This commit implements bidirectional translation of the llvm.uadd.with.overflow and the IAddCarry intrinsic.
Intrinsic llvm.uadd.with.overflow returns struct which second element have a type of i1.
The llvm type i1 is, in llvm-spirv, directly translated to BoolType.
SPIRV specification requires that the composite which returns from IAddCarry needs to have both elements of the same type.
In result, current implementation is not compliant and should be considered temporary.

### CODE REVIEW

File: lib/SPIRV/libSPIRV/SPIRVOpCode.h
Comment (asudarsa) [inline lib/SPIRV/libSPIRV/SPIRVOpCode.h:73]: I am wondering why this cannot be: (unsigned)OpCode >= OpIAdd && (unsigned)OpCode <= OpSMulExtended)
Ref: https://github.com/KhronosGroup/SPIRV-Headers/blob/79743b899fde5c954897b2694291002626358fac/include/spirv/unified1/spirv.h#L1377

File: test/llvm-intrinsics/uadd.with.overflow.ll
Comment (asudarsa) [inline test/llvm-intrinsics/uadd.with.overflow.ll:10]: May be we can add tests to make sure the SPIR-V types are correctly generated?
https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpIAddCarry

File: lib/SPIRV/SPIRVWriter.cpp
Comment (asudarsa) [inline lib/SPIRV/SPIRVWriter.cpp:3973]: Normal transType will not work for result here.
For example: if we have LLVM result type {i16, i1}, we are going to get {TypeInt, TypeBool} for SPIR-V. Instead, we need {TypeInt, TypeInt} according to spec: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpIAddCarry

The test is also failing in CI (spirv-val) because of this. Please take a look.

Thanks

File: lib/SPIRV/SPIRVReader.cpp
Comment (asudarsa) [inline lib/SPIRV/SPIRVReader.cpp:2498]: Please look at comment in SPIRVWriter.cpp. In the reverse translation, the second LLVM type will always be i1.

Thanks

File: lib/SPIRV/libSPIRV/SPIRVOpCode.h
Comment (bwlodarcz) [reply lib/SPIRV/libSPIRV/SPIRVOpCode.h:73]: The problem here is that I'm not sure what this function name actually means.
From context we can assume that binary in this case means two registers operator. If this is true than it means that this function doesn't cover all possible ops like e.g OpFOrdNotEqual.

File: test/llvm-intrinsics/uadd.with.overflow.ll
Comment (bwlodarcz) [reply test/llvm-intrinsics/uadd.with.overflow.ll]: Done

File: lib/SPIRV/SPIRVWriter.cpp
Comment (bwlodarcz) [reply lib/SPIRV/SPIRVWriter.cpp:3973]: Done

File: lib/SPIRV/SPIRVReader.cpp
Comment (bwlodarcz) [reply lib/SPIRV/SPIRVReader.cpp:2498]: Done

**Review by MrSidims — APPROVED** [review]
LGTM, will merge after @asudarsa approval

**Review by LU-JOHN — APPROVED** [review]
LGTM

File: lib/SPIRV/SPIRVReader.cpp
Comment (asudarsa) [inline lib/SPIRV/SPIRVReader.cpp:2501]: This type needs to be i1 in the resulting LLVM IR. 
Can you please check?

Thanks

Comment (bwlodarcz) [reply lib/SPIRV/SPIRVReader.cpp:2501]: This conversion already works by implicit conversion somewhere inside CreateBinaryIntrinsic. I'm not very fond of this kind of implicit behaviors and imo this should by default explode but it isn't. And now is a problem because without going inside CreateBinaryIntrinsic and finding out how conversion works there is no way to prove that forcing type of getOperand(1) works.
```
case OpIAddCarry: {
    IRBuilder Builder(BB);
    auto *BC = static_cast<SPIRVBinary *>(BV);
    SPIRVTypeBool BOC;
    BC->getOperand(1)->setType(&BOC);
    return mapValue(BV, Builder.CreateBinaryIntrinsic(
                            Intrinsic::uadd_with_overflow,
                            transValue(BC->getOperand(0), F, BB),
                            transValue(BC->getOperand(1), F, BB)));
  }
```
This would be a possible way to force type change.

**Review by asudarsa — APPROVED** [review]
LGTM. Please track this non-conformance with Spec by opening a github issue.

Thanks

**Pull request merged**
